sdk/python/feast/offline_store.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2019 The Feast Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Dict, List, Optional
+
+import pyarrow
+
+from feast.entity import Entity
+from feast.feature import Feature
+from feast.feature_view import FeatureView
+
+
+class OfflineStore(ABC):
+    """
+    OfflineStore is a non-user-facing object used for all interaction between Feast and the service used for offline storage of features. Currently BigQuery is supported.
+    """
+
+    @abstractmethod
+    def pull_table(
+        self, feature_view: FeatureView, start_date: datetime, end_date: datetime,
+    ):
+        pass
+
+
+class BigQueryOfflineStore:
+    """
+    BigQueryOfflineStore is a non-user-facing object used for all interaction between Feast and BigQuery.
+    """
+
+    def pull_table(
```

Comment:
`pull_table` seems like it squats generic functionality, but it's actually very specific (it contains a FeatureView argument and time range). I would expect this method to have a query interface like `pull_table(self, query)`. That way I can also reuse this method for generating a training dataset. I also think the method can be cleaner if the query building was done separately in a helper method.

Would it make sense to have two higher level methods for us in that case? Something like `export_feature_view_to_arrow(self, feature_view: FeatureView, start_date: datetime, end_date: datetime,
) -> Optional[pyarrow.Table]`
and 
`export_training_features_to_arrow(self, entity_df_query, features, feature_views) -> Optional[pyarrow.Table]`
and both of these would use
`_export_bigquery_table_to_arrow(self, query) -> Optional[pyarrow.Table] `

Reply:
changed the interface - would this be reusable for historical feature retrieval now?

Reply:
Feedback on the entity query was been quite positive. I dont think we can get away with asking users to provide a table for the entity df, even if they can write a view. It's a poor development experience, unless we provide other functionality like row filters, column filters, sampling, etc.

Reply:
Oh not saying to get rid of the entity query, the question was meant to ask is this method reusable in historical feature retrieval as a wrapper around the BQ storage api? The `export_training_features` method would still need to run the query outside of this method (and the BQ storage api doesn't support arbitrary queries anyway) to produce the table/view

Reply:
I'll try to make it work, but I am not sure about the table_ref. I can change it later if necessary though.